### ISSUE TITLE
JDBC instrumentation: classloader issue

### ISSUE THREAD

**Issue by claussenj** [open-telemetry/opentelemetry-java-instrumentation#6015]
**Describe the bug**
Classes related to JDBC instrumentation (e.g. io.opentelemetry.javaagent.shaded.instrumentation.jdbc.internal.AutoValue_DbInfo, io.opentelemetry.javaagent.shaded.instrumentation.jdbc.internal.DbInfo) are loaded into the app class loader, not into the agent class loader. This seems to contradict to the intended classloader separation (https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/docs/contributing/javaagent-jar-components.md). In addition this causes ClassCastExceptions since depending on the application the DbInfo and AutoValud_DbInfo classes arrive in different classloaders.

My request would be to clean this up and load all classes into the AgentClassLoader as documented.

**Steps to reproduce**
One way to look at it (or just use a debugger):
1. Instrument an application using JDBC with the opentelemetry agent. 
2. Let the application run a few SQL statements
3. Create a heap dump (e.g. jcmd <pid> GC.heap_dump) dump.hprof
4. Use Eclipse Memory Analyzer to open the HProf. Use "Java Basics" -> Class Loader Browser to get list of classes loaded by each classloader. Example screenshot and output attached.

**What did you expect to see?**
I would expect the JDBC instrumentation related classes to be loaded into the agent class loader.

**What did you see instead?**
The jdbc instrumentation classes are distributed over different application class loaders.

**What version are you using?**
1.13.1

**Environment**
Compiler: Sapmachine 11.0.14.1
OS: Red Hat 8.4

**Additional context**
My concrete test application is based on Spring Boot 2.6.7. Distribution of the JDBC instrumentation classes varies depending on how the JDBC driver is hooked into the class path: via -cp or into BOOT-INF/lib. This distribution triggers the actual ClassCastException. But I believe that the real fix would really be to load the instrumentation classes into the agent class loader.

Example exception:
java.lang.ClassCastException: class io.opentelemetry.javaagent.shaded.instrumentation.jdbc.internal.AutoValue_DbInfo cannot be cast to class io.opentelemetry.javaagent.shaded.instrumentation.jdbc.internal.DbInfo (io.opentelemetry.javaagent.shaded.instrumentation.jdbc.internal.AutoValue_DbInfo is in unnamed module of loader 'app'; io.opentelemetry.javaagent.shaded.instrumentation.jdbc.internal.DbInfo is in unnamed module of loader org.springframework.boot.loader.LaunchedURLClassLoader @3c9c6245)

Daemon Thread [http-nio-8082-exec-1] (Suspended (exception java.lang.ClassCastException))	
	owns: org.apache.tomcat.util.net.NioEndpoint$NioSocketWrapper  (id=140)	
	io.opentelemetry.javaagent.shaded.instrumentation.jdbc.internal.JdbcUtils.extractDbInfo(java.sql.Connection) line: 77	
	io.opentelemetry.javaagent.shaded.instrumentation.jdbc.internal.DbRequest.create(java.sql.Statement, java.lang.String) line: 36	
	io.opentelemetry.javaagent.shaded.instrumentation.jdbc.internal.DbRequest.create(java.sql.PreparedStatement) line: 26	
	com.zaxxer.hikari.pool.HikariProxyPreparedStatement.executeQuery() line: not available	

[Screenshots_MemoryAnalyzer.pdf](https://github.com/open-telemetry/opentelemetry-java-instrumentation/files/8676713/Screenshots_MemoryAnalyzer.pdf)
[ClassLoaderBrowser_export.txt](https://github.com/open-telemetry/opentelemetry-java-instrumentation/files/8676613/ClassLoaderBrowser_export.txt)

regards
Jens

**Comment by laurit** [comment]
See https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/docs/contributing/writing-instrumentation-module.md the classes you are seeing are referred as helper classes in that doc. The problem here is probably that these helper classes are injected into multiple class loaders in the same hierarchy and somehow one ends up being used where the other is expected.
Could you provide a small sample application that reproduces the issue, it would help a lot.

**Comment by claussenj** [comment]
Hi Lauri,
I guess you are right about the helper classes. Was not aware that they are intended to arrive in the application classloaders. Still the classloader conflict is blocking us.
I can try assembling an application to reproduce. Will probably take some time...
regards
Jens

**Comment by claussenj** [comment]
Hi Lauri,
I have found a way to reproduce the issue using spring-petclinic (https://github.com/spring-projects/spring-petclinic):

Procedure that I used:
* Setup Spring Petclinic as instructed in the README. I have used a mysql container, but the problem also occurs using a SAP HANA jdbc driver, so I assume that the problem is db-independent.
* put the otel agent and the jdbc driver in dedicated folders, e.g. OTELAGENT and jdbc
* Extract the fat Spring jar that results from the maven build and use the `org.springframework.boot.loader.JarLauncher` as given below. It seems to be important to reproduce 
1. that the Jarlauncher is used (and not java -jar petclinic.jar)
2. that the jdbc driver is listed explicitly via -cp

```
java -javaagent:../OTELAGENT/opentelemetry-javaagent-1.13.1.jar -Dotel.service.name=petclinic-service -Dotel.traces.exporter=none -Dotel.javaagent.enabled=true -agentlib:jdwp=transport=dt_socket,address=*:8000,server=y,suspend=n -Dspring.profiles.active=mysql -Dserver.port=9999 -cp ../jdbc/mysql-connector-java-8.0.29.jar:. org.springframework.boot.loader.JarLauncher
```

* connect with a debugger to the application
* set an exception breakpoint on "ClassCastException"
* access the petclinic , e.g. http://localhost:9999

Hope this helps to reproduce the problem.

regards
Jens

**Comment by system-designer** [comment]
> See https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/docs/contributing/writing-instrumentation-module.md the classes you are seeing are referred as helper classes in that doc. The problem here is probably that these helper classes are injected into multiple class loaders in the same hierarchy and somehow one ends up being used where the other is expected. Could you provide a small sample application that reproduces the issue, it would help a lot.

Thanks for your suggugestion,  I fix the problem by adding code below in AgentClassLoader:

    protected Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundException {
      // prometheus exporter uses jdk http server, load it from the platform class loader
      // my exporter uses jdk naming dns, load it from the platform class loader
      if (name != null
          && (name.startsWith("com.sun.net.httpserver.") || name.startsWith("com.sun.jndi.dns."))) {
        return platformClassLoader.loadClass(name);
      }
      return Class.forName(name, false, null);
    }

And then adding jvm args "--add-exports jdk.naming.dns/com.sun.jndi.dns=java.naming" to the test application.

**Comment by CMNiedermayer** [comment]
I'm in the same team like Jens and I can confirm that this issue could be resolved by applying OTEL agent version 1.15.0.

**Comment by claussenj** [comment]
Just for completeness a final comment from my side since I opened the issue and see that I have not responded:
Issue completely resolved for us in OTEL 1.15.0. Many thanks for the fix!
regards
Jens
